model/infinispan/src/main/java/org/keycloak/crl/infinispan/InfinispanCrlStorageProvider.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.crl.infinispan;
+
+import java.security.GeneralSecurityException;
+import java.security.cert.X509CRL;
+import java.util.Date;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.TimeUnit;
+import org.infinispan.Cache;
+import org.jboss.logging.Logger;
+import org.keycloak.common.util.Time;
+import org.keycloak.crl.CrlStorageProvider;
+
+/**
+ *
+ * @author rmartinc
+ */
+public class InfinispanCrlStorageProvider implements CrlStorageProvider {
+
+    private static final Logger log = Logger.getLogger(InfinispanCrlStorageProvider.class);
+
+    private final Cache<String, X509CRLEntry> crlCache;
+    private final Map<String, FutureTask<X509CRL>> tasksInProgress;
+    private final long cacheTime;
+    private final long minTimeBetweenRequests;
+
+    public InfinispanCrlStorageProvider(Cache<String, X509CRLEntry> crlCache, Map<String, FutureTask<X509CRL>> tasksInProgress,
+            int cacheTime, int minTimeBetweenRequests) {
+        this.crlCache = crlCache;
+        this.tasksInProgress = tasksInProgress;
+        this.cacheTime = cacheTime > 0? TimeUnit.SECONDS.toMillis(cacheTime) : -1;
+        this.minTimeBetweenRequests = minTimeBetweenRequests > 0? TimeUnit.SECONDS.toMillis(minTimeBetweenRequests) : 10000L;
+    }
+
+    @Override
+    public X509CRL get(String key, Callable<X509CRL> loader) throws GeneralSecurityException {
+        final X509CRLEntry crlEntry = crlCache.get(key);
+        final long currentTime = Time.currentTimeMillis();
+        if (crlEntry != null && (crlEntry.getCrl().getNextUpdate() == null || crlEntry.getCrl().getNextUpdate().compareTo(new Date(currentTime)) > 0)) {
+            log.debugf("returning CRL '%s' from cache because it's cached OK", key);
+            return crlEntry.getCrl();
+        }
+
+        // refresh the crl entry in the cache
+        return reloadCrl(key, loader, currentTime, crlEntry);
+    }
+
+    @Override
+    public boolean refreshCache(String key, Callable<X509CRL> loader) throws GeneralSecurityException {
```

Comment:
`refreshCache` is not used. Assuming it will be used in the future?

Reply:
I just did it because it is done in keys. In keys it is used to manually refresh the cache in the admin UI. I have maintained it but we can remove it for now if you prefer to not have it.

Reply:
No preference from me.